### ISSUE TITLE
Some Android devices can not find the 'gdx' shared library and the app crashes on launch

### ISSUE THREAD

**Issue by prineside** [libgdx/libgdx#7449]
#### Issue details
App crashes on launch on some devices, because 'libgdx.so' can not be found.
I've re-downloaded all of the natives (libgdx.so) and they are stored in libs/ (arm64-v8a, armeabi-v7a, x86 and x86_64).

The resulting .aab contains libgdx.so in such places, alongside with some other libraries:
- base/lib/arm64-v8a/
- base/lib/armeabi-v7a/
- base/lib/x86/
- base/lib/x86_64/

.apk builds also contain libgdx.so in such folders:
- lib/arm64-v8a/
- lib/armeabi-v7a/
- lib/x86/
- lib/x86_64/

The amount of crashes has spiked in the last update of my game, which probably means that something broke during Jul 27 - Aug 27.
Library loads properly on Google Pixel 4 XL / Android 13.

Some of the affected devices:
- Huawei P40 Pro 5G
- Samsung Galaxy S22 Ultra
- Xiaomi Redmi Note 10 Pro
- Vivo X9s
- Other devices by Google, Oneplus and others

Android versions 7-13.

Gradle config does not exclude specific ABIs and has not been changed in this last update. Crash reports mention both arm64 and x86_64 ABIs, but other architectures may also be affected. First thing I've noticed is that lib/arm64-v8a/ is on the list of native library divs but lib/arm64/ is mentioned in the exception, but then the same crash happened on x86_64 which should exist in lib/.
I think it may be somehow related to the updated dependency for jnigen-2.5.2

#### Reproduction steps/code
Can not reproduce this issue since I do not have an affected device model on hands. 
Only Firebase / Google Play Developer crash reports can confirm this.

#### Version of libGDX and/or relevant dependencies
GDX 1.12.2-SNAPSHOT
gdx-jnigen-loader-2.5.2 (SharedLibraryLoader.java)

#### Stacktrace
```java
com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: Couldn't load shared library 'gdx' for target: Android
    com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:174)
    com.badlogic.gdx.utils.GdxNativesLoader.load(GdxNativesLoader.java:30)
    com.badlogic.gdx.backends.android.AndroidApplicationConfiguration$1.load(AndroidApplicationConfiguration.java:0)
    com.badlogic.gdx.backends.android.AndroidApplication.init(AndroidApplication.java:121)
    com.badlogic.gdx.backends.android.AndroidApplication.initialize(AndroidApplication.java:87)
    com.prineside.tdi2.AndroidLauncher.onCreate(AndroidLauncher.java:468)
    android.app.Activity.performCreate(Activity.java:8241)
    android.app.Activity.performCreate(Activity.java:8207)
    android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
    android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3736)
    android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3937)
    android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
    android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
    android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
    android.app.ActivityThread$H.handleMessage(ActivityThread.java:2288)
    android.os.Handler.dispatchMessage(Handler.java:106)
    android.os.Looper.loopOnce(Looper.java:210)
    android.os.Looper.loop(Looper.java:299)
    android.app.ActivityThread.main(ActivityThread.java:8293)
    java.lang.reflect.Method.invoke(Method.java:0)
    com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:556)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: library "libgdx.so" not found
    java.lang.Runtime.loadLibrary0(Runtime.java:1077)
    java.lang.Runtime.loadLibrary0(Runtime.java:998)
    java.lang.System.loadLibrary(System.java:1656)
    com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:168)
    com.badlogic.gdx.utils.GdxNativesLoader.load(GdxNativesLoader.java:30)
    com.badlogic.gdx.backends.android.AndroidApplicationConfiguration$1.load(AndroidApplicationConfiguration.java:0)
    com.badlogic.gdx.backends.android.AndroidApplication.init(AndroidApplication.java:121)
    com.badlogic.gdx.backends.android.AndroidApplication.initialize(AndroidApplication.java:87)
    com.prineside.tdi2.AndroidLauncher.onCreate(AndroidLauncher.java:468)
```
Also some of the devices with an older version of Android have a slightly different exception message:
```
Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/com.prineside.tdi2-QrLU1wkDtvtULNebxu5F3w==/base.apk"],nativeLibraryDirectories=[/data/app/com.prineside.tdi2-QrLU1wkDtvtULNebxu5F3w==/lib/x86_64, /system/lib64, /system/vendor/lib64]]] couldn't find "libgdx.so"
```
or
```
Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.prineside.tdi2-z2Z6yXu2E4HmC4clowIzHA==/base.apk"],nativeLibraryDirectories=[/data/app/com.prineside.tdi2-z2Z6yXu2E4HmC4clowIzHA==/lib/arm64, /system/lib64, /hw_product/lib64, /system/product/lib64]]] couldn't find "libgdx.so"
```
I can send an .apk file if that helps. The affected version of the game is also currently available on Google Play.

#### Please select the affected platforms
- [x] Android
- [ ] iOS
- [ ] HTML/GWT
- [ ] Windows
- [ ] Linux
- [ ] macOS

**Comment by badlogic** [comment]
Can you tell me on what Google devices it crashes (hopefully including the OS version)? Then I might be able to reproduce it with one of my Google phones.

**Comment by prineside** [comment]
Google Pixel 6 Pro, Android 12
Google Nexus 5, Android 7.1.2
Google devices are not that popular in my case and make ~7% of the affected devices

**Comment by prineside** [comment]
May be unrelated but I've also started to receive a weird crash reports from a piece of code which is only executed if ```Gdx.input.isKeyJustPressed(Input.Keys.F11)```, meaning this method returns true for some reason (on Android). This happens mostly on devices like TECNO, Infinix, DOOGEE, Blackview and Itel (Mediatek / Spreadtrum SoCs), and I would blame the manufacturer, but it is weird this issue has only emerged in the last update

**Comment by prineside** [comment]
And another fresh weird crash has emerged in this last update:
```
backtrace:
  #00  pc 0x0000000000519800  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!54f602101825a16f797fc1a29cdd4f!194d5e5755!+4968)
  #01  pc 0x0000000000268c0c  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!ae9df5840a3b187aa9f98544704c1f!194d5e5755!+12124)
  #02  pc 0x000000000025eb94  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!d0051afa8fb0bb02a68e169738355f!194d5e5755!+492)
  #03  pc 0x00000000002c9c2c  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!94050ffb92ce9d2eb541a2faeb7ea2!194d5e5755!+156)
  #04  pc 0x00000000002c9a0c  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!10f39af3d5dda97a64124dee6a3783!194d5e5755!+132)
  #05  pc 0x00000000003dd734  /vendor/lib64/egl/libGLESv2_adreno.so (!!!0000!04fe7c7713a0660472688458ecb7a2!194d5e5755!+148)
  #06  pc 0x0000000000032e14  /system/lib64/libEGL.so (android::eglSwapBuffersWithDamageKHRImpl(void*, void*, int*, int)+608)
  #07  pc 0x000000000002f370  /system/lib64/libEGL.so (eglSwapBuffers+56)
  #08  pc 0x00000000000f4c58  /system/lib64/libandroid_runtime.so (android::jni_eglSwapBuffers(_JNIEnv*, _jobject*, _jobject*, _jobject*)+108)
  #09  pc 0x0000000000352140  /data/misc/apexdata/com.android.art/dalvik-cache/arm64/boot.oat (art_jni_trampoline+144)
  #10  pc 0x00000000005ba854  /apex/com.android.art/lib64/libart.so (nterp_helper+7636)
  #11  pc 0x00000000001d2348  /system/framework/framework.jar (android.opengl.GLSurfaceView$EglHelper.swap+12)
  #12  pc 0x00000000005b99d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924)
  #13  pc 0x00000000001d2d48  /system/framework/framework.jar (android.opengl.GLSurfaceView$GLThread.guardedRun+1152)
  #14  pc 0x00000000005b99d4  /apex/com.android.art/lib64/libart.so (nterp_helper+3924)
  #15  pc 0x00000000001d3354  /system/framework/framework.jar (android.opengl.GLSurfaceView$GLThread.run+52)
  #16  pc 0x000000000033b3a4  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #17  pc 0x000000000023a5d0  /apex/com.android.art/lib64/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+144)
  #18  pc 0x0000000000539bf4  /apex/com.android.art/lib64/libart.so (art::Thread::CreateCallback(void*)+1600)
  #19  pc 0x00000000005395a4  /apex/com.android.art/lib64/libart.so (art::Thread::CreateCallbackWithUffdGc(void*)+8)
  #20  pc 0x00000000000c37b4  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+208)
  #21  pc 0x000000000005d084  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68)
  ```
  This one also affects Galaxy S22/S23 and may be related somehow

**Comment by badlogic** [comment]
Someone on Discord brought up that users may have pirated your APK, and
such pirated APKs often lead to weird crashes in the dev console in Google
Play (likely because them pirates mix in their own evil code). We might be
looking at such a situation here. Your latest update may have brought your
app to the attention of more users who downloaded the APK from some site.
Can you tell us what app it is? I think I have a Pixel 6 Pro somewhere here.

**Comment by prineside** [comment]
https://play.google.com/store/apps/details?id=com.prineside.tdi2&hl=en
A modified game may indeed be the reason, but unfortunately I can not confirm nor deny it. I'll try to search for a modified versions of the game on the Internet for now

**Comment by badlogic** [comment]
Sweet game! Works fine on my Pixel 6 Pro. The theory above may have some
merit (other theory being people flushing some very terrible ROMs onto
their devices)

**Comment by prineside** [comment]
Thank you for testing it!
I'll try to find a bit more info on this topic and will probably add a couple of integrity checks with the next hotfix update.
It may take a couple of days for the update to be validated and published, I'll comment on this thread as soon as I find something.

**Comment by prineside** [comment]
Update:
24 after another update, everything (mostly) looks good.
I've only received 4 crash reports related to libgdx.so, coming from 3 different users, 2 of which look like some kind of a Google's pre-release testing devices (Google's IP, "Generic x86-x64 device") and the last one has failed the signature check. I've also disabled Crashlytics (Firebase) reporting if the app detects signature mismatch, so it may still happen on those devices we do not care about.
libgdx.so files for all platforms were changed somehow in this last update, according to my local VCS - maybe it has affected the result.
Another observation - "nativeLibraryDirectories" paths contain strings like "/data/app/com.prineside.tdi2-1/...", "/data/app/com.prineside.tdi2-HgeMtxQ2L8owbkHK2tT8HA==/..." and "/data/app/com.prineside.tdi2-QrLU1wkDtvtULNebxu5F3w==" (base64 decodes into some binary stuff) while, according to my Pixel 4 device running Android 13, app directory does not contain that suffixes when installed properly. First I thought that this is just some kind of a security measure which makes it harder for the malicious apps to access the files of other apps, but that one "-1" case makes me think it is just a copy of the application made by 3rd party software (LuckyPatcher or something). Signature check has not failed, probably because the signature itself has been included as a regular String, which is easy to patch.

TL;DR; Probably this issue is related to the modified versions of the game, you were right. I'll try to obfuscate the signature, to make it harder to patch the game, and will keep an eye on the crash reports related to this issue.

Thank you! Closing this issue

**Comment by prineside** [comment]
2 updates later, I've gathered some info.
App signature is correct, files are not modified, package name is correct but crash reports are still coming.

I'm suspecting it may be caused by a software which makes a copy of an app and fails to properly do so, but can't confirm it is used here because I can't find an answer to these questions:
- Is it possible to make a copy of some app and keep exactly the same package name for it / to make a copy of an app with a different package name and make the app see the correct package name? All reported crashes have a correct package name.
- Are app installation directories obfuscated somehow? Does the vanilla Android OS always install apps into a clean path like "/data/app/[package_name]/" or maybe it appends some secret string in order to make it harder for the other apps to access these files directly / to scan the file system? 

I've got crash reports from 16 different users in 2 days after release, from Samsung, Huawei, Oppo and other devices, Android 8-12.

The most suspicious part is the app's installation directory, some examples:
- /data/app/com.prineside.tdi2-kLNP6ut5vcuKhILe1lA-TA==/base.apk
- /data/app/~~sGhA9l8emXX7JdGSZnHApA==/com.prineside.tdi2-2Cl9rrfL42fGUFtUyhLMoQ==/base.apk
- /data/app/~~rMOckeTOxaWExqcw1FkeMg==/com.prineside.tdi2-hR2cbCiUy6JmnrprDI7ifg==/base.apk
- /data/app/com.prineside.tdi2-GGNfjGXPb1pssLv5_-LZjQ==/base.apk
- /data/app/com.prineside.tdi2-dlEDtz9DbCyCiMx4_2g4tw==/base.apk
And none of the reports have a "/data/app/com.prineside.tdi2/" installation dir. 

But I can't make a conclusion, because I've also gathered this info from the users who do not have issues loading libgdx.so and, assuming it is very unlikely their app is modified, this installation dir format seems to be forced by Android itself. Some examples from devices not experiencing the shared library loading issue:
- /data/app/~~orB5qCuYA5vRGuKDQfd_BQ==/com.prineside.tdi2-QJ0dmgxLbSyZ9OTw2QnW0A==/base.apk
- /data/app/~~ln2w925dPvFGj1h5JPyH7Q==/com.prineside.tdi2-iNdgh3g5P12ILmNQuEU6PQ==/base.apk
- /data/app/~~z8-bXimo7OvcyFNSx-aHgw==/com.prineside.tdi2-_SJ0xRjuRPt66JdQWNF51g==/base.apk
- /data/app/~~0gokAGCmt1qy93sL7kPq1A==/com.prineside.tdi2-D2bQx9t-0JlM3z5EV1nusA==/base.apk
- /data/app/com.prineside.tdi2-ZpGARqciIK58iTHWQLSK0Q==/base.apk
- /data/app/~~9Dow9_PJp0FwWvOFAqg4hg==/com.prineside.tdi2--aF7Iv-61FQ2nAc4QhP4Dg==/base.apk
...

In short:
- App package is correct (which makes me believe it is not a copy of the app)
- App installation dir looks legit
- File integrity checks passed
- App signature is valid (using an obfuscated implementation of a signature check to make it hard for a 3rd party software to patch it day 1 after release)

**Comment by Frosty-J** [comment]
> Is it possible to make a copy of some app and keep exactly the same package name for it

Yes; it's more work to change the package name than to not change it, since you'd have to modify the APK to change it.. For example, if I share an app with someone over Bluetooth, it will have the same package name and all for them as it did for me. This process has its drawbacks these days, as my installation might not have the native libraries, drawable densities, or language resources best suited for their device.

**Comment by prineside** [comment]
Thanks for the reply, it totally makes sense now.

Bluetooth... or 3rd party websites sharing an .apk for only some specific architecture (just realized that). This explains day 1 reports - those websites must be monitoring Google Play and extracting .apk automatically and quickly. Same package name, proper signature, but different architecture.

I'll add some checks and will try to contact the affected players, closing the issue for now because it will take some time for me to release another update and to gather more info.
